atom/browser/ui/win/toast_lib.cc

```diff
@@ -0,0 +1,537 @@
+// copies or substantial portions of the Software.
+
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+#include "toast_lib.h"
+#include "browser/win/scoped_hstring.h"
+#pragma comment(lib,"shlwapi")
+#pragma comment(lib,"user32")
+
+using namespace WinToastLib;
+namespace DllImporter {
+
+    // Function load a function from library
+    template <typename Function>
+    HRESULT loadFunctionFromLibrary(HINSTANCE library, LPCSTR name, Function &func) {
+        if (!library)
+            return false;
+
+        func = reinterpret_cast<Function>(GetProcAddress(library, name));
+        return (func != nullptr) ? S_OK : E_FAIL;
+    }
+
+    typedef HRESULT(FAR STDAPICALLTYPE *f_SetCurrentProcessExplicitAppUserModelID)(__in PCWSTR AppID);
+    typedef HRESULT(FAR STDAPICALLTYPE *f_PropVariantToString)(_In_ REFPROPVARIANT propvar, _Out_writes_(cch) PWSTR psz, _In_ UINT cch);
+    typedef HRESULT(FAR STDAPICALLTYPE *f_RoGetActivationFactory)(_In_ HSTRING activatableClassId, _In_ REFIID iid, _COM_Outptr_ void ** factory);
+    typedef HRESULT(FAR STDAPICALLTYPE *f_WindowsCreateStringReference)(_In_reads_opt_(length + 1) PCWSTR sourceString, UINT32 length, _Out_ HSTRING_HEADER * hstringHeader, _Outptr_result_maybenull_ _Result_nullonfailure_ HSTRING * string);
+    typedef HRESULT(FAR STDAPICALLTYPE *f_WindowsDeleteString)(_In_opt_ HSTRING string);
+
+    f_SetCurrentProcessExplicitAppUserModelID    SetCurrentProcessExplicitAppUserModelID;
+    f_PropVariantToString                        PropVariantToString;
+    f_RoGetActivationFactory                     RoGetActivationFactory;
+    f_WindowsCreateStringReference               WindowsCreateStringReference;
+    f_WindowsDeleteString                        WindowsDeleteString;
+
+
+    template<class T>
+    _Check_return_ __inline HRESULT _1_GetActivationFactory(_In_ HSTRING activatableClassId, _COM_Outptr_ T** factory) {
+        return RoGetActivationFactory(activatableClassId, IID_INS_ARGS(factory));
+    }
+
+    template<typename T>
+    inline HRESULT Wrap_GetActivationFactory(_In_ HSTRING activatableClassId, _Inout_ Details::ComPtrRef<T> factory) throw() {
+        return _1_GetActivationFactory(activatableClassId, factory.ReleaseAndGetAddressOf());
+    }
+
+    inline HRESULT initialize() {
+        HINSTANCE LibShell32 = LoadLibrary(L"SHELL32.DLL");
+        HRESULT hr = loadFunctionFromLibrary(LibShell32, "SetCurrentProcessExplicitAppUserModelID", SetCurrentProcessExplicitAppUserModelID);
```

Comment:
@felixrieseberg Hmm, now that I think about it maybe not.  `app.setUserModelID` probably handles this for us, will test on Windows 8 (where appUserModelID is strictly required)